File: PoseKit/Sources/PoseKit/ReferencePose.swift
#if canImport(Foundation)
import Foundation
#endif

/// An authored target pose bundled with the app as JSON.
public struct ReferencePose: Codable, Identifiable, Equatable, Sendable {
    public let id: String
    public let title: String
    public let tags: [String]
    /// Intent-collection ids this pose belongs to (e.g. "dating", "mirror").
    public let collections: [String]
    /// Free tier: usable without a Pose+ subscription.
    public let free: Bool
    public let joints: [String: [Float]]

    public init(id: String, title: String, tags: [String],
                collections: [String] = [], free: Bool = false,
                joints: [String: [Float]]) {
        self.id = id
        self.title = title
        self.tags = tags
        self.collections = collections
        self.free = free
        self.joints = joints
    }

    private enum CodingKeys: String, CodingKey {
        case id, title, tags, collections, free, joints
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        title = try c.decode(String.self, forKey: .title)
        tags = try c.decode([String].self, forKey: .tags)
        collections = try c.decodeIfPresent([String].self, forKey: .collections) ?? []
        free = try c.decodeIfPresent(Bool.self, forKey: .free) ?? false
        joints = try c.decode([String: [Float]].self, forKey: .joints)
    }

    /// Typed pose. Unknown joint names and arrays without exactly 2 values are skipped.
    public var poseVector: PoseVector {
        var pts: [Joint: SIMD2<Float>] = [:]
        for (key, arr) in joints {
            guard let joint = Joint(rawValue: key), arr.count == 2 else { continue }
            pts[joint] = SIMD2<Float>(arr[0], arr[1])
        }
        return PoseVector(points: pts)
    }
}

File: android/posekit/src/main/kotlin/com/oerol/posekit/ReferencePose.kt
package com.oerol.posekit

import org.json.JSONObject

/**
 * An authored target pose bundled with the app as JSON. Parsed with org.json
 * (part of the Android framework; a plain jar on the JVM) so the library adds
 * no serialization dependency.
 */
data class ReferencePose(
    val id: String,
    val title: String,
    val tags: List<String>,
    val collections: List<String>,
    val free: Boolean,
    val joints: Map<String, List<Float>>,
) {
    /** Typed pose. Unknown joint names and arrays without exactly 2 values are skipped. */
    val poseVector: PoseVector
        get() {
            val pts = mutableMapOf<Joint, Vec2>()
            for ((key, arr) in joints) {
                val joint = Joint.fromKey(key) ?: continue
                if (arr.size != 2) continue
                pts[joint] = Vec2(arr[0], arr[1])
            }
            return PoseVector(pts)
        }

    companion object {
        /** Decodes the same JSON schema the iOS app bundles. */
        fun fromJson(json: String): ReferencePose {
            val obj = JSONObject(json)
            val tagsArr = obj.getJSONArray("tags")
            val tags = (0 until tagsArr.length()).map { tagsArr.getString(it) }
            val jointsObj = obj.getJSONObject("joints")
            val joints = mutableMapOf<String, List<Float>>()
            for (key in jointsObj.keys()) {
                val arr = jointsObj.getJSONArray(key)
                joints[key] = (0 until arr.length()).map { arr.getDouble(it).toFloat() }
            }
            val collectionsArr = obj.optJSONArray("collections")
            val collections = if (collectionsArr == null) emptyList()
                else (0 until collectionsArr.length()).map { collectionsArr.getString(it) }
            val free = obj.optBoolean("free", false)
            return ReferencePose(
                id = obj.getString("id"),
                title = obj.getString("title"),
                tags = tags,
                collections = collections,
                free = free,
                joints = joints,
            )
        }
    }
}
